cleverhans/model.py

```diff
@@ -0,0 +1,107 @@
+        # The key is a pair of input (the symbolic representation of the input)
+        # and train (a boolean indicating whether the graph is in training or
+        # inference mode---which changes the behavior of layers like dropout).
+        # The values output_dict are also a dictionary mapping layer names
+        # to symbolic representation of the output of that layer.
+        self.fprop_cache = {}
+
+        # By default, we assume the model is being used for training (i.e.,
+        # 'train' time). If the model is used for inference or in a different
+        # state, a call to set_state() should be made first.
+        self.state = 'train'
+
+        pass
+
+    def __call__(self, *args, **kwargs):
+        """
+        For compatibility with functions used as model definitions (taking
+        an input tensor and returning the tensor giving the output
+        of the model on that input).
+        """
+        return self.get_probs(*args, **kwargs)
+
+    def set_state(self, state):
+        """
+        Set which state the model is currently being used in. This can be any
+        string and is used to by the cache to keep different output tensors
+        for each of the states used. The default value of self.state is 'test'
+        as a reference to the inference phase, but you could set it as 'train'
+        (which would for instance mean that the model uses dropout layers).
+        :param state: (string) A name to identify which state the model is in.
+        """
+        self.state = state
+        print('The model state was set to: ' + state)
+        return
+
+    def get_layer(self, x, layer):
+        """
+        Expose the hidden features of a model given a layer name.
+        :param x: A symbolic representation of the network input
+        :param layer: The name of the hidden layer to return features at.
+        :return: A symbolic representation of the hidden features
+        """
+        # Return the symbolic representation for this layer.
+        return self.fprop(x)[layer]
+
+    def get_logits(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output logits (i.e., the
+                 values fed as inputs to the softmax layer).
+        """
+        raise NotImplementedError('`get_logits` not implemented')
+
+    def get_probs(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output probabilities (i.e.,
+                the output values produced by the softmax layer).
+        """
+        raise NotImplementedError('`get_probs` not implemented')
+
+    def get_layer_names(self):
+        """
+        :return: a list of names for the layers that can be exposed by this
+        model abstraction.
+        """
+        raise NotImplementedError('`get_layer_names` not implemented')
+
+    def fprop(self, x):
```

Comment:
Is there a particular use case where it would be hard to make sure the logits layer is named "logits"? It seems trivial to me but maybe I'm not being creative enough thinking of corner cases.